### PR TITLE
Getting removing and adding list's WIP

### PR DESCRIPTION
**Pull request by whiteadam2**

### CODE REVIEW

File: apps/data-scanner/lib/api/scanner/app.mjs [hunk 1]
```diff
@@ -29,11 +30,17 @@ async function scanStatuses(config) {
...
             // todo: check for new sessions addSessionsList
             // todo: sessions.add(addSessionsList)
+
             // todo: add functionality that is checking phone numbers that should be added or removed
             // todo: read DB and return two lists (addList, removeList)
+            const phonesToAdd = await new Phones().getAddingList();
+            console.log(phonesToAdd);
+            const phonesToRemove = await new Phones().getRemovingList();
+            console.log(phonesToRemove);
+
```
Comment (yzhbankov) [inline apps/data-scanner/lib/api/scanner/app.mjs:43]: cleanup logs

File: apps/data-scanner/lib/db/migrations/sqls/20220421082245-alter-sessions-table-up.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+ALTER TABLE sessions 
+    ADD COLUMN valid BOOLEAN DEFAULT TRUE, 
+    ADD COLUMN fulled BOOLEAN DEFAULT FALSE;
```
Comment (yzhbankov) [inline apps/data-scanner/lib/db/migrations/sqls/20220421082245-alter-sessions-table-up.sql:3]: filled?

File: apps/data-scanner/lib/api/scanner/app.mjs [hunk 3]
```diff
@@ -29,11 +30,17 @@ async function scanStatuses(config) {
     const job = new CronJob(
         config.cron,
         async function () {
-            await sessions.invokeEach(new Api.contacts.GetContacts({}));
+            //  await sessions.invokeEach(new Api.contacts.GetContacts({}));
```
Comment (yzhbankov) [inline apps/data-scanner/lib/api/scanner/app.mjs:33]: uncomment

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     * @property updateSession
+     */
+
+    /**
+     * @method
+     * @returns {Promise<Object>}
+     */
+    async getAddingList() {
+        const phones = await this.repository.phones.read({ tracked: true });
+        return phones.filter((phone) => !phone.session_id);
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/phones/Phones.mjs:17]: implement filtering on sql level. pass one more parameter to read method({ tracked: true, session: true })

File: apps/data-scanner/lib/api/scanner/app.mjs [hunk 5]
```diff
@@ -29,11 +30,17 @@ async function scanStatuses(config) {
...
             // todo: check for new sessions addSessionsList
             // todo: sessions.add(addSessionsList)
+
             // todo: add functionality that is checking phone numbers that should be added or removed
             // todo: read DB and return two lists (addList, removeList)
+            const phonesToAdd = await new Phones().getAddingList();
```
Comment (yzhbankov) [inline apps/data-scanner/lib/api/scanner/app.mjs:39]: const inclusionPhonesList = new Phones().getList({ tracked: true, session: false });
const exclusionPhonesList = new Phones().getList({ tracked: false, session: true });

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    /**
+     * @method
+     * @param {String} trackedPhoneId - tracked phone identifier
+     * @param {String} sessionId - session identifier
+     * @returns {Promise<Object>}
+     */
+    async updateSession(params) {
+        return this.repository.session.update(params);
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/phones/Phones.mjs:36]: do not mix different repos in one model

File: shared/repository/lib/api/documents/PhonesRepository.mjs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+        super(db);
+    }
+
+    /**
+     * @method
+     * @param {boolean} tracked - the flag that indicates do the phone is tracking
+     * @return {Promise<Object>} returns data saved in DB
+     */
+    async read({ tracked }) {
```
Comment (yzhbankov) [inline shared/repository/lib/api/documents/PhonesRepository.mjs:24]: { tracked, session }

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+     * @property getAddingList
+     * @property getRemovingList
+     * @property updateSession
+     */
+
+    /**
+     * @method
+     * @returns {Promise<Object>}
+     */
+    async getAddingList() {
```
[the remark this answers is not in the document]
Comment (whiteadam2) [reply apps/data-scanner/lib/models/phones/Phones.mjs:15]: done

File: apps/data-scanner/lib/api/scanner/app.mjs (hunk 5 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/api/scanner/app.mjs:39]: done

File: shared/repository/lib/api/documents/PhonesRepository.mjs (hunk 7 above)
Comment (whiteadam2) [reply shared/repository/lib/api/documents/PhonesRepository.mjs:24]: done

File: apps/data-scanner/lib/models/phones/Phones.mjs (hunk 4 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/models/phones/Phones.mjs:17]: done

File: apps/data-scanner/lib/api/scanner/app.mjs (hunk 3 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/api/scanner/app.mjs:33]: done

File: apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs [hunk 9]
```diff
@@ -17,11 +17,11 @@ export class UpdateSessions {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs:17]: phone --> sessionId

File: apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs [hunk 10]
```diff
@@ -17,11 +17,11 @@ export class UpdateSessions {
 
             const session = new Session();
 
-            const sessionByPhone = await session.readByPhone({ phone });
+            const sessionByPhone = await session.readBySessionId({ sessionId: phone });
```
Comment (yzhbankov) [inline apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs:20]: sessionByPhone --> sessionData (rawSession)
await (new Session()).readBySessionId

File: apps/data-scanner/lib/db/migrations/sqls/20220410070449-add-table-sessions-up.sql [hunk 11]
```diff
@@ -14,5 +14,4 @@ CREATE TABLE sessions(
     dc_id TEXT NOT NULL,
     server_address INET NOT NULL,
     port INT NOT NULL,
-    active BOOLEAN NOT NULL DEFAULT TRUE
```
Comment (yzhbankov) [inline apps/data-scanner/lib/db/migrations/sqls/20220410070449-add-table-sessions-up.sql:17]: revert

File: apps/data-scanner/lib/db/migrations/sqls/20220421082245-alter-sessions-table-up.sql [hunk 12]
```diff
@@ -0,0 +1,6 @@
+ALTER TABLE sessions 
+    ADD COLUMN valid BOOLEAN DEFAULT TRUE, 
+    ADD COLUMN is_full BOOLEAN DEFAULT FALSE;
+
```
Comment (yzhbankov) [inline apps/data-scanner/lib/db/migrations/sqls/20220421082245-alter-sessions-table-up.sql:4]: delete active field in sessions table

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import ModelBase from '../ModelBase.mjs';
+
+export class Phones extends ModelBase {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/phones/Phones.mjs:3]: TrackedPhone

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+        return await this.repository.phones.read({ tracked, withSession });
+    }
+
+    /**
+     * @method
+     * @param {String} trackedPhone - tracked phonenumber which session we should update
+     * @param {String} sessionId - session identifier
+     * @returns {Promise<Object>}
+     */
+    async updateSession(params) {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/phones/Phones.mjs:26]: updateSession ---> update

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+export class Phones extends ModelBase {
+    /**
+     * @typedef {Class} Phones
+     * @property getList
+     * @property updateSession
+     */
+
+    /**
+     * @method
+     * @param {String} tracked - flag for phone numbers which should be added to traking
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/phones/Phones.mjs:12]: String --> Boolean

File: apps/data-scanner/lib/models/phones/Phones.mjs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+     * @param {String} withSession - flag for phone numbers which have session field filled
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/phones/Phones.mjs:13]: String --> Boolean

File: apps/data-scanner/lib/models/sessions/Client.mjs [hunk 17]
```diff
@@ -29,7 +29,7 @@ export class Client {
      * @param {String} apiConfig.apiHash - Telegram Api hash
      */
     constructor(sessionConfig, apiConfig) {
-        this.#uid = uuidv4();// todo: replce uid with sessionId
+        this.#uid = uuidv4(); // todo: replce uid with sessionId
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Client.mjs:32]: this.#uid  --> this.#sessionId
not use uuid

File: apps/data-scanner/lib/models/sessions/Session.mjs [hunk 18]
```diff
@@ -3,37 +3,51 @@ import ModelBase from '../ModelBase.mjs';
...
 
     /**
      * @method
-     * @param {String} phone - Phone nubmer that activates the session
+     * @param {String} sessionId - session idetifier that equel phone number that activates the session
      * @returns {Promise<Object>}
      */
-    async readByPhone({ phone }) {
-        return this.repository.session.readByPhone({ phone });
+    async readBySessionId({ sessionId }) {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Session.mjs:17]: readBySessionId --> readById

File: apps/data-scanner/lib/models/sessions/Session.mjs [hunk 19]
```diff
@@ -3,37 +3,51 @@ import ModelBase from '../ModelBase.mjs';
...
+        return this.repository.session.readBySessionId({ sessionId });
     }
 
     /**
      * @method
+     * @param {String} sessionId - session idetifier that equel phone number that activates the session
+     * @param {Boolean} valid - if session is valid
+     * @param {Boolean} full - if session is full of phone numbers
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Session.mjs:25]: full --> isFull change in code everywere

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 20]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
      * @returns {Promise<void>}
      */
     async init() {
-        const sessions = await new Session().readAll();
+        const sessions = await new Session().readAll({ valid: true });
+
+        this.#pool['full'] = [];
+        this.#pool['avaliable'] = [];
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:44]: this.#pool['full'] = {};
this.#pool['avaliable'] = {};

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 21]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
...
+     * @returns {Promise<Object>}
+     */
+    async addPhones(phones, attempts = 2) {
+        const addedPhones = [];
+        let result = null;
+        let errors = 0;
+
+        for (let phone of phones) {
+            do {
+                result = await addOnePhone(phone);
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:111]: this.addOnePhone

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 22]
```diff
@@ -16,9 +19,9 @@ export class Sessions {
     #config = null;
 
     /**
-     * @property {Client[]}
+     * @property {}
      */
-    #pool = [];
+    #pool = {};
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:24]: #poll = {
  available: new Map(),
  filled: new Map(),
}

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 23]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
...
+        for (let session of sessions) {
+            const client = await new Client(session, this.#config);
             await client.init();
-            this.#pool.push(client); // todo: replace array with object { sessionId: client }
-            // [{sessionId, api }, {sessionId, api }, {sessionId, api }]
-            // { 
-            //     1234: {sessionId:1234, api },
-            //     3456: {sessionId:3456, api },
-            // }
+            if (session.is_full) this.#pool['full'].push({ sessionId: session.id, client });
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:49]: this.#pool.filled.set(session.id, { sessionId: session.id, client });
else
this.#pool.available.set(session.id, { sessionId: session.id, client });

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 24]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
...
+    // todo: mark in DB as expired
     // }
 
     // #findSessionById(sessionId) {}
 
+    async addOnePhone(phoneNumber) {
+        const avaliablePool = this.#pool['avaliable'];
+        let result = null;
+
+        if (avaliablePool.length === 0) {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:69]: if (this.#pool.available.size)

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 25]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
...
+            console.log('There are no avaliable sessions!');
+            return false;
+        }
+
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:73]: const iterator = this.#pool.available.entries();

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 26]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
...
     // async del() {
-        // todo: implement logic to remove session to the pool
-        // todo: mark in DB as expired
+    // todo: implement logic to remove session to the pool
+    // todo: mark in DB as expired
     // }
 
     // #findSessionById(sessionId) {}
 
+    async addOnePhone(phoneNumber) {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:65]: addOnePhone --> addPhone

File: apps/data-scanner/lib/models/sessions/Sessions.mjs [hunk 27]
```diff
@@ -35,31 +38,86 @@ export class Sessions {
...
+        let errors = 0;
+
+        for (let phone of phones) {
+            do {
+                result = await addOnePhone(phone);
+                errors++;
+            } while (!result.addedPhone || errors < attempts);
+
+            if (result.addedPhone) addedPhones.push(result);
+            else await new Session().update({ sessionId: result.sessionId, full: false, valid: true });
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/Sessions.mjs:116]: const session = new Session().read(sessionId);
session.filled = true;
new Session().save(session);

new Session().update({ sessionId: result.sessionId, full: false, valid: true });

update(params) {
const parms = [];
const values = []
 if (params.filled) {
  params.push('is_full')
  values.push(params.filled)
}

sql

File: apps/data-scanner/lib/models/phones/Phones.mjs (hunk 14 above)
Comment (yzhbankov) [reply apps/data-scanner/lib/models/phones/Phones.mjs:26]: const rawPhone = await this.repository.phones.readByPhone(phone);
rawPhone.sessionId = params.sessionId;
return this.repository.phones.update(params);

File: apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs (hunk 9 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs:17]: done

File: apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs (hunk 10 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/api/scanner/usecases/UpdateSessions.mjs:20]: done

File: apps/data-scanner/lib/db/migrations/sqls/20220410070449-add-table-sessions-up.sql (hunk 11 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/db/migrations/sqls/20220410070449-add-table-sessions-up.sql:17]: done

File: apps/data-scanner/lib/db/migrations/sqls/20220421082245-alter-sessions-table-up.sql (hunk 12 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/db/migrations/sqls/20220421082245-alter-sessions-table-up.sql:4]: done

File: shared/repository/lib/api/documents/ContactRepository.mjs [hunk 28]
```diff
@@ -18,37 +19,29 @@ export class UserPhonesRepository extends RepoBase {
...
+            id,
+            user_id: userId,
+            tracked_phone: trackedPhone,
+            tracked,
+            created_at: createdAt,
+            updated_at: updatedAt,
+            session_id: sessionId,
+        };
+
+        return await dbRead(this.db, 'tracked_phones', dbProps);
```
Comment (yzhbankov) [inline shared/repository/lib/api/documents/ContactRepository.mjs:43]: do not use a very common method for selecting from DB. it is hard to debug and read

File: shared/repository/lib/api/documents/ContactRepository.mjs [hunk 29]
```diff
@@ -18,37 +19,29 @@ export class UserPhonesRepository extends RepoBase {
...
-     */
-         async readOne({ userId, phone }) {
-            const sql = 'SELECT * FROM tracked_phones WHERE user_id = $1 AND tracked_phone = $2';
-            const params = [userId, phone];
-            const result = await this.db.queryAsync(sql, params).catch((err) => {
-                throw new RepoError(err);
-            });
-            return result && result.rows[0];
-        }
+    async read({ id, userId, trackedPhone, tracked, createdAt, updatedAt, sessionId }) {
```
Comment (yzhbankov) [inline shared/repository/lib/api/documents/ContactRepository.mjs:32]: why do you pass all these parameters to read?

File: apps/data-scanner/lib/models/sessions/ContactTracked.mjs [hunk 30]
```diff
@@ -1,8 +1,8 @@
 import ModelBase from '../ModelBase.mjs';
 
-export class Phones extends ModelBase {
+export class ContactTracked extends ModelBase {
```
Comment (yzhbankov) [inline apps/data-scanner/lib/models/sessions/ContactTracked.mjs:3]: ContactTracked --> UserContact

File: shared/repository/lib/api/documents/ContactRepository.mjs [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+    async read({ userId, tracked, trackedPhone, sessionId, withSession }) {
+        let sql = 'SELECT * FROM tracked_phones';
+        let params = '';
+
+        if (typeof userId === 'number') {
+            params += ` user_id = ${userId}`;
+        }
+
+        if (typeof tracked === 'boolean') {
+            params !== '' ? (params += ` AND tracked = ${tracked}`) : (params += ` tracked = ${tracked}`);
```
Comment (yzhbankov) [inline shared/repository/lib/api/documents/ContactRepository.mjs:52]: params += params ? ' AND' : '';
params += ` tracked = ${tracked}`;

File: shared/repository/lib/api/documents/ContactRepository.mjs [hunk 32]
```diff
@@ -0,0 +1,150 @@
...
+
+    /**
+     * @method
+     * @param {Number} userId - user identifier
+     * @param {String} trackedPhone - contact added by user that should be tracked
+     * @param {Boolean} tracked - The flag that indicates is the contact tracked
+     * @returns {Promise<Object>}
+     */
+
+    async update({ userId, trackedPhone, tracked }) {
```
Comment (yzhbankov) [inline shared/repository/lib/api/documents/ContactRepository.mjs:104]: I thought that we agreed that it should be all properties list

File: shared/repository/lib/api/documents/ContactRepository.mjs (hunk 31 above)
Comment (whiteadam2) [reply shared/repository/lib/api/documents/ContactRepository.mjs:52]: done

File: shared/repository/lib/api/documents/ContactRepository.mjs (hunk 32 above)
Comment (whiteadam2) [reply shared/repository/lib/api/documents/ContactRepository.mjs:104]: done

File: shared/repository/lib/api/documents/ContactRepository.mjs [hunk 33]
```diff
@@ -0,0 +1,150 @@
...
+        return result && result.rows;
+    }
+
+    /**
+     * @method
+     * @param {Number} userId - user identifier
+     * @param {String} trackedPhone - contact added by user that should be tracked
+     * @returns {Promise<Object>}
+     */
+    async save({ userId, trackedPhone }) {
```
[the remark this answers is not in the document]
Comment (whiteadam2) [reply shared/repository/lib/api/documents/ContactRepository.mjs:87]: agreed that no

File: shared/repository/lib/api/documents/ContactRepository.mjs [hunk 34]
```diff
@@ -0,0 +1,150 @@
...
+     * @param {string} trackedPhone - contact added by user that should be tracked
+     * @param {string} sessionId - identifier of session that tracks this contact
+     * @param {boolean} withSession - flag that indicates does contact have session
+     * @return {Promise<Object>} returns data saved in DB
+     */
+    async read({ userId, tracked, trackedPhone, sessionId, withSession }) {
+        let sql = 'SELECT * FROM tracked_phones';
+        let params = '';
+
+        if (typeof userId === 'number') {
```
[the remark this answers is not in the document]
Comment (whiteadam2) [reply shared/repository/lib/api/documents/ContactRepository.mjs:47]: it's couldnt. it has numeric type on db lavel

File: apps/data-scanner/lib/models/phones/Phones.mjs (hunk 14 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/models/phones/Phones.mjs:26]: done

File: apps/data-scanner/lib/models/phones/Phones.mjs (hunk 16 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/models/phones/Phones.mjs:13]: done

File: apps/data-scanner/lib/models/sessions/Session.mjs (hunk 19 above)
Comment (whiteadam2) [reply apps/data-scanner/lib/models/sessions/Session.mjs:25]: done

**Review by yzhbankov — APPROVED** [review]

**Pull request merged**